### ISSUE TITLE
Atomicity of space:drop()

### ISSUE THREAD

**Issue by mkostoevr** [tarantool/tarantool#8667]
**Bug description**

[box.schema.space.drop](https://github.com/tarantool/tarantool/blob/246e8232b68a4abfa85910532d8ad8ff93d72741/src/box/lua/schema.lua#L882) can leave the database in an inconsistent state if one of its steps failed.

* OS: Linux
* OS Version: Ubuntu 22.04.2 LTS
* Architecture: amd64

Tarantool version: [current HEAD](https://github.com/tarantool/tarantool/commit/246e8232b68a4abfa85910532d8ad8ff93d72741) (as of 17 may 2023).

**Steps to reproduce**

1. Configure the tarantool box: `box.cfg{}`.
2. Attempt to drop `_vpriv` system space (will fail): `box.space._vpriv:drop()`.
3. Attempt to drop `_vinyl_deferred_delete` (should fail by the same reason, but... see below): `box.space._vinyl_deferred_delete:drop()`.

**Actual behavior**

```
tarantool> box.space._vpriv:drop()
---
- error: Can't drop the primary key in a system space, space '_vpriv'
...

tarantool> box.space._vinyl_deferred_delete:drop()
---
- error: 'builtin/box/schema.lua:230: attempt to index field ''object'' (a nil value)'
...
```

**Expected behavior**

The first error is expected, but the second one is not.

**Root cause**

1. On attempt to drop the `_vpriv` we drop all of its indexes starting from non-primary ones here: [box.schema.space.drop+25](https://github.com/tarantool/tarantool/blob/246e8232b68a4abfa85910532d8ad8ff93d72741/src/box/lua/schema.lua#L907), so the space has no indexes anymore.
2. Then we try to get its `object` index on `_vinyl_deferred_delete` drop here: [box.schema.space.drop+27](https://github.com/tarantool/tarantool/blob/246e8232b68a4abfa85910532d8ad8ff93d72741/src/box/lua/schema.lua#L909)->[revoke_object_privs+3](https://github.com/tarantool/tarantool/blob/246e8232b68a4abfa85910532d8ad8ff93d72741/src/box/lua/schema.lua#L230). But since we have deleted all non-primary indexes of the `_vpriv` before, tarantool fails with nil-access error.

**Proposal**

The `box.schema.space.drop` function should revert all the changes it made if any step is failed.

**Comment by olegrok** [comment]
Seems it's related to https://github.com/tarantool/tarantool/issues/4348

In case if drop will be atomic we have following case
```lua
tarantool> box.atomic(function() box.space._vpriv:drop() end)
---
- error: Can't drop the primary key in a system space, space '_vpriv'
...

tarantool> box.atomic(function() box.space._vinyl_deferred_delete:drop() end)
---
...
```

**Comment by mkostoevr** [comment]
> Seems it's related to #4348

Yes, it is.

> In case if drop will be atomic we have following case
> 
> ```lua
> tarantool> box.atomic(function() box.space._vpriv:drop() end)
> ---
> - error: Can't drop the primary key in a system space, space '_vpriv'
> ...
> 
> tarantool> box.atomic(function() box.space._vinyl_deferred_delete:drop() end)
> ---
> ...
> ```

Yes, and the behavior is desirable. Now you can drop the `_vinyl_deferred_delete` space, but it will be fixed if the related PR (#8623) will be merged.

Closed the issue as duplicate of #4348.
